src/com/ms/silverking/cloud/dht/daemon/storage/StorageModule.java

```diff
@@ -150,425 +200,545 @@ public StorageModule(NodeRingMaster2 ringMaster, String dhtName, Timer timer, Zo
+          metricsObserver.initialize(metricsNamespaceStores);
+          Log.info("Initialized StorageMetricsObserver: " + metricsObserverName);
+        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+          Log.logErrorWarning(e, "Unable to instantiate StorageMetricsObserver: " + metricsObserverName);
         }
+      }
     }
-    
-    private void startLinkWatches() {
-    	for (NamespaceStore nsStore: namespaces.values()) {
-            nsStore.startWatches(zk, nsLinkBasePath, this);                	
-    	}
+  }
+
+  public void stop() {
+    cleanerTask.cancel();
+    reapTask.cancel();
+
+    methodCallNonBlockingWorker.stopLWTPool();
+    methodCallBlockingWorker.stopLWTPool();
+    if (nsMetaStore != null) {
+      nsMetaStore.stop();
     }
+    namespaces.clear();
+    zk.close();
 
-    private void recoverExistingNamespace(File nsDir) throws IOException {
-        try {
-            long    ns;
-            NamespaceProperties nsProperties;
-            NamespaceStore  parent;
-            NamespaceStore	nsStore;
-                
-            Log.warning("\t\tRecovering: "+ nsDir.getName());
-            ns = NumConversion.parseHexStringAsUnsignedLong(nsDir.getName());
-            nsProperties = NamespacePropertiesIO.read(nsDir);
-            nsMetaStore.setNamespaceProperties(ns, nsProperties);
-            if (nsProperties.getParent() != null) {
-                long    parentContext;
-                
-                parentContext = new SimpleNamespaceCreator().createNamespace(nsProperties.getParent()).contextAsLong();
-                parent = namespaces.get(parentContext);
-                if (parent == null) {
-                    throw new RuntimeException("Unexpected parent not found: "+ parentContext);
-                }
-            } else {
-                parent = null;
-            }
-            nsStore = NamespaceStore.recoverExisting(ns, nsDir, parent, null, mgBase, ringMaster, 
-                    activeRetrievals, zk, nsLinkBasePath, this, reapPolicy);
-            namespaces.put(ns, nsStore);
-            nsStore.startWatches(zk, nsLinkBasePath, this);            
-            Log.warning("\t\tDone recovering: "+ nsDir.getName());
-        } catch (NumberFormatException nfe) {
-            nfe.printStackTrace();
-            Log.warning("Recovery ignoring unexpected nsDir: ", nsDir);
-        }
+  }
+
+  private void createMetaNSStore() {
+    long metaNS;
+
+    metaNS = NamespaceUtil.metaInfoNamespace.contextAsLong();
+    if (namespaces.get(metaNS) == null) {
+      NamespaceStore metaNSStore;
+
+      metaNSStore = new NamespaceStore(metaNS, new File(baseDir, NamespaceUtil.contextToDirName(metaNS)),
+          NamespaceStore.DirCreationMode.CreateNSDir, NamespaceUtil.metaInfoNamespaceProperties,
+          //spGroup.getRootPolicy(),
+          mgBase, ringMaster, false, activeRetrievals);
+      namespaces.put(metaNS, metaNSStore);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **299 and 306**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).